fx quant: add a BC test for loading old torch.package models

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65538

Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31512939](https://our.internmc.facebook.com/intern/diff/D31512939)

vkuzo committed:
fx quant: add a BC test for loading old torch.package models

Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/bc/test_backward_compatibility.py:
```diff
@@ -42,8 +48,25 @@ def get_filenames(self, subname):
+        traced_module_file, expected_file, package_file, get_attr_targets_file
+
+
+class LinearReluFunctional(nn.Module):
```
For the model it might be useful to introduce some kind of a hierarchy, to make sure the attr FQNs correctly reflect that. Similar to https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L2151

vkuzo replied:
sure

supriyar commented on test/quantization/bc/test_backward_compatibility.py:
```diff
@@ -142,6 +167,66 @@ def _test_obs(self, obs, input_size, subname=None, generate=False, check_numeric
+        imp = torch.package.PackageImporter(package_file)
+        m = imp.load_pickle(package_name, resource_name_model)
+        mq = _do_quant_transforms(m, input_tensor)
+        get_attrs = _get_get_attr_target_strings(mq)
```
Can we also get the `prepared.named_buffers()` and compare the observer buffers in the saved model vs the new one?

vkuzo replied:
sure

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar reviewed (approved):
Looks good, please take a look at the failing test before landing.

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31512939).

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31512939](https://our.internmc.facebook.com/intern/diff/D31512939)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31512939).

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31512939](https://our.internmc.facebook.com/intern/diff/D31512939)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31512939).

vkuzo committed:
Update on "fx quant: add a BC test for loading old torch.package models"


Summary:

Adds a test which verifies that `prepare_fx` and `convert_fx` work
on models created by `torch.package` in the past.  In detail:

1. (one time) create a model and save it with torch.package. Also save input,
expected output, and names of quantization related get_attrs added by
our passes.
2. (every time) load the model from (1), and verify that expected output
matches current output, and that get_attr targets did not change.

Test Plan:

```
python test/test_quantization.py TestSerialization.test_linear_relu_package_quantization_transforms
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31512939](https://our.internmc.facebook.com/intern/diff/D31512939)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31512939).
